### ISSUE TITLE
Support ES6 and beyond?

### ISSUE THREAD

**Issue by michaelficarra** [purescript/purescript#2207]
Inspired by #2204 (which I strongly support), I think we can avoid waiting for language-javascript to support modern JavaScript features by instead implementing a simple lexer. I don't believe we actually require the AST that language-javascript gives us. If we just look for the token sequences that indicate an export, we can infer that the name is exported regardless of context (thus not requiring a full parse). And if I'm not mistaken, the only reason we parse the FFI is to confirm that the required names are exported.

I'd be willing to do this work. Thoughts? Concerns?

**edit:** I forgot to mention, but this also allows adoption of future ECMAScript features in FFI without explicit support in PureScript, language-javascript, or some dependency. For instance, TypeScript/Flow annotations would pass through just fine.

**Comment by michaelficarra** [comment]
I forgot to mention, but this also allows adoption of future ECMAScript features in FFI without explicit support in PureScript, language-javascript, or some dependency. For instance, TypeScript/Flow annotations would pass through just fine.

**Comment by hdgarrood** [comment]
I think we do need to know where any given export assignment statement ends for DCE in `psc-bundle`, which presumably would require some amount of actual parsing, rather than just lexing. It's perhaps still arguable that we don't need a full-blown parser to do this, but I'd personally rather depend on a more widely-used and separately released package than a not-quite-parser that only we use.

**Comment by michaelficarra** [comment]
@hdgarrood I was thinking that DCE (and possibly all of psc-bundle) might be unnecessary after #2204 and related. We already have tools that do this for sufficiently declarative JS.

**Comment by hdgarrood** [comment]
Good point. Did you perhaps mean to link to #2206 though?

**Comment by hdgarrood** [comment]
Also, I am 90% sure that you are indeed correct - that in `psc`, we only parse FFI to ensure that the required names match up with what's in the PureScript module. How complicated do you think a JS lexer would be?

**Comment by michaelficarra** [comment]
@hdgarrood Yes, "and related" was meant to include #2206 and any similar issues that are spawned from #2204.

> Also, I am 90% sure that you are indeed correct

That's about how sure I am as well. Confirmation from @garyb would be helpful here.

> How complicated do you think a JS lexer would be?

Well I have a significant amount of experience working on JS parsers, so I think I have a pretty good idea how difficult a standalone JS lexer will be, but I'm not sure how I can convey that to you. Are you looking for number of lines of code or something?

**Comment by garyb** [comment]
I'm also only 90% sure! @hdgarrood wasn't it you who added the FFI parsing to check exports? I didn't think we did any FFI parsing before that, aside from the ad-hoc find `// module` comment thing.

`language-javascript` was initially added only for `psc-bundle` if I remember rightly.

**Comment by hdgarrood** [comment]
It was me who added it, yes, and it was the only part of `psc` that parsed FFI modules when I added it, but I'm not 100% certain that we haven't added more code since then that does other stuff with it.

@michaelficarra I was aware of your experience :) I guess I'm asking how easily someone else might be able to read through the code and fix a bug, were we to find one.

If we go for ES6 modules and try to reuse existing JS tooling as much as possible, which I think is very probably going to happen at some point anyway, we might also consider dropping this feature from `psc`, and making build tools responsible for integrating something to do this check if people still want it.

**Comment by hdgarrood** [comment]
For instance, if you were running a DCE tool, that would probably catch this anyway, right?

**Comment by hdgarrood** [comment]
("this" being an error where you `foreign import` something that isn't exported by the relevant module.)

**Comment by michaelficarra** [comment]
@hdgarrood That is a very good point. It does seem like an awful lot of effort to go through to do something that an external tool is probably going to be doing anyway.

**Comment by texastoland** [comment]
@michaelficarra Trying this together with #1206 unless someone else is interested.

**Comment by michaelficarra** [comment]
The recent PR https://github.com/purescript/purescript/pull/2210 gives a reason to continue finding FFI definitions. psc-ide-server can return the definition location of a value. This is helpful for IDE jump-to-definition functionality.

**Comment by texastoland** [comment]
With a custom lexer though? It's a bummer we can't use Babel or TS for FFI because of `language-javascript`.

**Comment by michaelficarra** [comment]
> With a custom lexer that just identifies exports though?

Yes. What's wrong with that? That's what I've proposed in this issue.

**Comment by texastoland** [comment]
Just confirming 💓

**Comment by dimsmol** [comment]
Getting back to the matters brought up by #2204 and #1206 (this issue is linked as a continuation of them), I would say that generating modern JS and relying on the tools from JS universe (webpack, babel, etc.) may look like a good idea, but in practice it isn't that good.

Both webpack and babel aren't really reliable, neither they are fast. We have a huge project that takes more than 4 minutes to build. We are systematically suffering from all kinds of the nasty bugs coming from both webpack and babel or one of their dependencies.

Every npm-shrinkwrap.json generation for a dependency update is a risk of destabilizing everything because a tiny bug in any of the myriads of dependencies and sub-dependencies used by webpack/babel can break something. With npm, there is no good way to prevent these dependencies from switching to higher (and probably broken) versions even if you are updating just some unrelated stuff in your package.json

Maybe webpack 2 is more stable, I don't know. But I doubt it is because it's really hard to write reliable software in JS and there is no good quality control for all those dependencies. Both webpack and babel have a lot of code full of that JS-style interlinked data bags being passed around, especially where it relates to plugins and loaders.

Modern JS is something that cannot work without further processing, and unfortunately there is no really robust tools for this. With all my respect to webpack and babel I'd try to avoid them if I could. However, my current project is JS mostly and I cannot do it anyway.

Direct ES5 generation is faster and more robust than ES6 generation with further processing by babel or any other transpiler. And it's much easier than writing our own robust ES6 to ES5 transpiler tool.

As for bundling, @wegry shared his experience in #2204. We have very similar case (huge JS code base with growing but still relatively small PureScript part). But we have very different results. In our case, using purs-loader with PureScript bundling and DCE reduces resulting code size very well comparing to using purs-loader without bundling.

The difference is that we avoid depending on JS code from PureScript as much as we can. This allows PureScript bundler to optimize PureScript part and webpack to optimize JS modules independently. As it was already said, JS tree shaking will hardly be as effective as PureScript-aware DCE (or will require "tree shaking-optimized" code generation which can be suboptimal in many other ways).

Modern JS is good, but it's also good to be able to avoid depending on webpack/babel. From the practical standpoint, it will be very nice to preserve support for ES5 + CommonJS code generation and PureScript-aware bundling we currently have.

**Comment by Pauan** [comment]
I agree with @dimsmol 

I also do JavaScript development with Webpack/Babel/Rollup. They work well enough for JS code, but I have had issues with breakage.

On a fundamental level, a PureScript bundler will be able to perform optimizations that a JS bundler will **never** be able to do. This is because of PureScript's static + pure behavior, compared to JavaScript's dynamic + impure behavior.

PureScript *will* need to support ECMAScript 6... just not right now. When ES6 becomes more mainstream (enabled by default in Node.js and/or modern browsers), then we will need to migrate over. But for now, I think the status quo is okay.

On the other hand, I think it would be good to grab some low-hanging fruit: it would be nice if PureScript would parse ES6 in FFI files, so that it would work with tools like `psc-bundle` and `psc-ide-server`.

In other words, rather than doing full ES6->ES5 transpilation, it would simply parse the `import` and `export` statements. This is similar to what Rollup does.

**Comment by Pauan** [comment]
I have the following concrete proposal:

1. Add in support to `psc-bundle` for bundling ES6 files. A file is detected as ES6 if it contains either `import` or `export`, otherwise it's assumed that it is CommonJS.

   It would not do any transpilation, so only `import`/`export` would be supported. This is just like how Rollup works. If you want transpilation you would need to use Babel, just like right now.

  `psc` would continue to output CommonJS and `psc-bundle` would continue to bundle CommonJS files. So this is a backwards-compatible change which *adds* ES6 module support, it does not *replace* CommonJS.

2. After Node.js and/or browsers support ES6 modules, then change `psc` to output ES6 modules.

   Because of step 1, `psc-bundle` already supports ES6 bundling, so it doesn't need to change.

   Existing FFI code doesn't need to change either, because `psc-bundle` supports both CommonJS *and* ES6 modules. So this should also be a backwards-compatible change.

   At this point we can also consider having `psc` output other ES6 features (e.g. `class`, `=>`, etc.)

3. Slowly phase out CommonJS support from `psc-bundle`. PureScript FFI code would need to migrate over to ES6 modules (if they haven't already).

Step 1 shouldn't take that long, it just requires somebody to do the work.

Steps 2 and 3 will probably take a few years, but it doesn't require much work from the PureScript compiler, it's just a general community transition.

**Comment by wegry** [comment]
Sorry for the wall of text.

> a PureScript bundler will be able to perform optimizations that a JS bundler will never be able to do

@dimsol Correct me if I'm wrong, but the optimizations #1206 covered (undocumented elsewhere?) in 0.9.x `psc-bundle` currently are: 

	- Live module loading (only what's used)
	- Tree shaking (which is more aggressive about output JS using things like `new` that have side effects)
	- Decurrying

As covered in #1206, there doesn't seem to be enough there currently to warrant usage in projects that happen to mix JS and PS.

I understand that the PS community finds JS and its tooling as whole undesirable. I've just been pushing for not crowding out alternate implementations and want no part in that debate.

For us, at least, `psa` wrapping `psc 0.8` sometimes takes 12 minutes or hangs for hours to do a full rebuild of 400 modules and 3 MB of CJS to 400 files. Webpack linting and transpiling 60 and loading 800 modules in 60 seconds is pretty much the cost of doing business. There's warts to go all around.

Decurrying is the only optimization my team currently dearly misses, because the other two issues are mostly covered by Webpack 2 + postprocessing. Decurrying could potentially be patched up with eval/apply instead of push/enter in psc-@ekmett might be able to answer questions there-but that's probably it's whole own issue.

The biggest drawbacks to using `psc-bundle` for us-also partially covered in #1206-were:

	- Only supports CJS when we have dependencies with UMD, AMD, and legacy exports
	- Leaking global variables
	- No async loading of code of non-essential code, a la webpack's chunking or AMD loading
	- Size and minification issues due to psc-bundle's storing modules as strings and export logic (40% reduction there)
	- Tree shaking JS dependencies

`psc`'s FFI export checker and artificial ES5 enforcement could be superceded by something like Webpack 2's ES module import checks which meet the static criteria laid out by @dimsol. I believe there was a discussion about dropping that, but as always, take that with a grain of salt.

In response to @Pauan's post:

	1. Rollup's biggest flaw currently is that it only supports one module style (setting aside plugins for now). Webpack is currently exploring ways to pull ideas from Rollup. `psc` is in the same boat with CommonJS. It's fine until you want to use code from another module system.

Before my team used ES modules, we used AMD and obviously CJS for PS. To get our `psc` bundled code loaded, at runtime mind you, we had to wrap the whole legacy module psc-bundle exports in an AMD module.

My team, in an enterprise environment, doesn't control the entire page. So periodically, someone would change a dependency version and stomp on our library with a different version or API in the global namespace. If someone has to use a fallen JS dependency that hangs variables on the window (jQuery, Lodash, etc.), how does psc-bundle account for that?

Webpack allowed us to, at the very least, shield our internal dependencies from the global namespace-common libraries like jQuery and Lodash. In addition to bundling and minifing, we also use it for placing and versioning svgs, fonts, images, LESS, etc.

It's not really an apples to apples comparison between psc-bundle and webpack when those things are considered. Juche is fine but it's not free. And I'm not advocating breaking up the old `pulp > psa > psc-bundle > psc` toolchain other teams use. It's about being accomodating to teams that use alternate toolchains. We use `webpack > psa > psc`.

psc, on account of only supporting valid ES identifiers as of 0.9, doesn't have to change anything to use [the loader I linked to very recently](https://github.com/purescript/purescript/issues/1206#issuecomment-249419838) in #1206 for tree-shaking. No transpilation required @dimsmol.

This has the added benefit of as @Pauan mentioned:

> Existing FFI code doesn't need to change

	2. ES6 features will be nice as far as getting code to the browser goes. Things like object shorthand and arrows being the biggest bang for your buck. Release Firefox is at 94%, but Chrome, Edge, and Safari are all at 100% support outside of modules.

	3. No comment ;)

**Comment by Pauan** [comment]
@wegry I agree that Webpack is amazing when you want to seamlessly tie into the rest of the JS ecosystem, including all the wacky ES6/CommonJS/RequireJS/npm/global stuff. I also agree that PureScript should have good support for Webpack.

In my opinion, good support for alternative workflows (e.g. Webpack) is a separate (but very important!) issue. However, there may be some overlap between the two issues.

----

I tried out your Webpack loader, but I had to make some minor tweaks to get it to work right:

https://gist.github.com/Pauan/53b1af11a71d6ebac8defafad070b3fe

It's obviously very hacky, but it does work: in my testing, it was able to remove **all** of the `require`, `exports`, and `module.exports` from both PureScript and FFI code.

Despite that, Webpack2 still didn't do a very good job of tree shaking. From what I understand, Webpack2 doesn't do any actual treeshaking itself, instead it relies on UglifyJS to remove dead code. But UglifyJS is *very* conservative in the code that it removes, so the end result is a lot of bloat. In addition to that, Webpack2 still wraps each ES6 module in a separate function, which also prevents tree shaking.

I will try to write a Rollup plugin and then we can see if Rollup does a better job of tree shaking compared to Webpack2.

**Comment by Pauan** [comment]
Okay, I created a repository which tests the different ways of doing tree shaking with PureScript:

https://github.com/Pauan/purescript-tree-shaking

It tests a program which is only slightly more complex than "Hello world". It imports JS files into PureScript, imports JS files into JS files, and imports PureScript files into JS.

I found that the best way to make it work is to use `pulp build` to output CommonJS files to the `output` folder and then afterwards slurp them up with Rollup/Webpack.

The `purs-loader` plugin works okay for Webpack, but I ran into some issues when trying to import an FFI file into another FFI file. If I'm totally abusing the FFI and I'm not supposed to do that, please tell me.

If anybody has any suggestions for improvements, you can either tell me or submit a pull request.

P.S. This feels a bit off-topic, so perhaps we should move this discussion to a new issue?

Correctness
---------------

* `pulp-browserify` fails to build, due to weird path issues when an FFI file requires another FFI file.

* `psc-bundle` builds successfully, but fails at runtime because it doesn't support bundling `require` inside of FFI files.

* `webpack` builds successfully, but fails at runtime with a bizarre error. In addition, the output looks really gnarly, so I suspect there is some sort of bug with the custom `purs-loader.js`

* `rollup` builds successfully, and it works correctly at runtime too! In fact it's the only one that works correctly. It also produces very clean code.

Size
-----

* `pulp-browserify` can't build, so N/A

* `psc-bundle` is `14.6 kB` regular, `6.9 kB` minified

* `rollup` is `30.8 kB` regular, `12.1 kB` minified

* `webpack` is `105.2 kB` regular, `24.4 kB` minified

I suspect that psc-bundle's dead code elimination is more aggressive than Rollup, probably due to purity.

Webpack's output is super bad, which makes me think that I'm doing something wrong. If anybody has any ideas, please do share.

**Comment by dimsmol** [comment]
@wegry thanks for explanation!
I totally agree, making PureScript to play nice with third-party bundling (webpack or any other tools) is an important task.

Speaking about your particular case, as I understand, everything already works via not using psc-bundle and do all the bundling with webpack instead. Do you have any problems with your workflow that must be addressed in PureScript tooling?

I guess, in your scenario you could benefit from PureScript DCE if it could work on it's own without bundling. It still won't be at it's full power for JS code importing PureScript modules because in this case it works on module level, but it can remove a lot of stuff within PureScript code. Tracking which PureScript modules are required can be tricky, purs-loader actually [does it incorrectly](https://github.com/ethul/purs-loader/issues/68) for now, but it can potentially be solved. If we could understand what exact parts of PureScript modules are required, PureScript DCE could work with all its power even for imports from JS code.

I propose to consider the following:
- Make it possible to do DCE for PureScript without bundling (have separate modules as a result, but with all the dead code eliminated).
- Allow to provide more fine-grained information for dead code eliminator - not only modules needed, but also particular exports required from those modules. With this feature it will be possible to improve, say, purs-loader to use PureScript DCE at it's max power.

For the drawbacks of psc-bundle you've mentioned, I totally agree that it would be nice to address the problems with global variables leakage and large module name strings. At the same time I don't think it's a task for psc-bundler to do JS tree-shaking or support wide variety of JS modules. It's better to use JS tools like webpack if you need such things. For async loading, it's a nice feature to have, but I'm ok with psc-bundler not having it (again, one can use webpack or some other JS bundler if really need it).

I like the proposition by @Pauan to incorporate ES6 modules support. In future we will need them anyway. I also like an idea to start using other ES6 features in compilation results, but not by default. Maybe it worth to have compilation flag(s) for this. Eventually we will need to use them all, but browsers and node.js must be ready first.

One more thing I want to address is psc's inability to understand non-CJS exports when checking FFI modules. I don't think there should be a support for wide variety of JS modules, but having support for ES6 would be nice because it's a future anyway. For the other kinds of modules the problem can be solved by making this check optional (allow to turn it off via compiler flag). The integrity in these cases must be checked by JS tooling (if possible).

The common idea is to make PureScript tools work best with the PureScript stuff and not care about JS specifics too much, but at the same time to not block the way for JS tools one may want to use over resulting JS code.

**Comment by wegry** [comment]
@pauan If you think you're "abusing" the FFI, my team has doen much worse. You can use `psc-bundle` with the FFI having external imports JS, you just need to ensure the require path in the FFI is relative to where the bundle will be, as opposed to the unbundled file. The import also can't be at the top level or `psc-bundle` will eat the require and do nothing. You will then need to use a module loader or bundler that understands that import after this.

And that's excellent to see to see the loader sort of working, thanks for taking it for a spin. There is an open issue about [rollup's perf relative to Webpack 2] (https://github.com/webpack/webpack/issues/2873#issuecomment-241458071) and plans on how to integrate it into Webpack 2.

One thing you might try with webpack in your sample repo is to run uglify through webpack's production mode which uses uglify under the covers, i.e. `webpack -p` which may fix your error. Uglify doesn't do tree shaking either and only supports ES5.

I brought up the custom loader because it, or something like it might make transitioning in the event the community decides that it wants to update from CJS to ES modules. It also makes PS exports statically analyzable for JS consumption.

Some tree shaking is better than none though, and output can always get better for targeting uglify and rollup/babel-transform. I may not have stressed enough that `psc-bundle` is currently much better for non-mixed projects.

@dimsmol The idea of a flag came up in #1206 for module type-I believe it was shot down. I'm personally in support of it. @texastoland has (or had) implemented a version of it already, if you'd like to talk to him about it.

> the same time I don't think it's a task for psc-bundle to do JS tree-shaking or support wide variety of JS modules

I absolutely agree that the amount of work to do this is probably prohibitive. Other tools can fill the gap.

**Comment by starper** [comment]
I don't know, maybe it's too radical, but have anyone thought about full `es6` support as a separate backend? And after some time `es5` backend could be marked as deprecated and `es6` backend (already tested by many users) could become the main one.

**Comment by Pauan** [comment]
@wegry With regard to using `require` with `psc-bundle`, you're right, but that doesn't work when you're trying to `require` an FFI file inside of a different FFI file: the FFI file ends up being included twice in the bundle! That's what I was referring to when I talked about "abusing the FFI".

I tried using `webpack -p` but I got the same result. However, when I was optimizing the Rollup plugin, I figured out what the problem is. It's fixed now, so Webpack also works. Here are the new sizes:

* `psc-bundle` is `14.6 kB` regular, `6.9 kB` minified

* `rollup` is `29.6 kB` regular, `11.3 kB` minified

* `webpack` is `105.0 kB` regular, `23.8 kB` minified

I think if Rollup's tree shaking can be improved, it will be on par with `psc-bundle`. It's already very close, it just needs to be able to tree shake ES5 "classes". I posted [a comment](https://github.com/rollup/rollup/issues/179#issuecomment-250903867), let's see how things go.

----

@starper Yes, it's been mentioned, but @paf31 doesn't want multiple backends (for good reason):

https://github.com/purescript/purescript/issues/1206#issuecomment-228821858

https://github.com/purescript/purescript/issues/1206#issuecomment-228824522

https://github.com/purescript/purescript/issues/1206#issuecomment-228923156

As much as I would love ES6 codegen, it's tricky to do it right.

----

Another idea that @paf31 had is to use a flag in `psc-bundle` to generate 1 CommonJS module per PureScript function: https://github.com/purescript/purescript/issues/1206#issuecomment-229705276

This could enable much better dead code elimination in bundlers like Webpack or Rollup. This doesn't solve every use case, but it has the benefit that it is easy to do and it's backwards-compatible, whereas adding in ES6 support is much harder.

**Comment by natefaubion** [comment]
> the FFI file ends up being included twice in the bundle!

I would definitely consider requiring FFI files from other FFI files an abuse. If you need lots of shared functionality then it probably best to put that functionality in another package, or relocate all you FFI bindings to a shared internal module (reexporting as needed). `psc` copie the FFI files around on disk (to your output directory), and writes the artifacts into a flat list keyed by module name. This obviously very likely changes the module identifier you'd use to require a sibling FFI file. If it still can magically resolve to the old source location, then you'd get multiple copies (the original source file, and the relocated compilation artifact).

**Comment by Pauan** [comment]
@natefaubion Indeed, that is how `psc` currently operates. I created an issue about that: https://github.com/purescript/purescript/issues/2343

**Comment by Pauan** [comment]
Hey guys, [my pull request](https://github.com/rollup/rollup/pull/1017) just got merged into Rollup, so I tried rerunning [my tree-shaking tests](https://github.com/Pauan/purescript-tree-shaking).

Here are the results:

* `psc-bundle` is `14.4 kB` regular, `6.8 kB` minified
* `rollup` is `10.8 kB` regular, `4.0 kB` minified
* `webpack` is `103.8 kB` regular, `23.7 kB` minified

Minor improvements for `psc-bundle` and Webpack, but **massive** improvements for Rollup. Rollup is now officially smaller than `psc-bundle`!

So I think it is a viable option to push more for Rollup, at least as an alternative to `psc-bundle`, similar to how Pulp supports Browserify.

I plan to eventually make a Rollup plugin for PureScript, but no guarantees.

**Comment by paf31** [comment]
@Pauan That's great news! Could you please share the resulting JS code in a gist so that we can see the difference?

This is pretty exciting, since it means there is the possibility of reducing the scope of `psc-bundle` quite a bit for 1.0.

**Comment by Pauan** [comment]
@paf31 [Done.](https://gist.github.com/Pauan/3af525619a0bdf1f670043e0d70e2770) I didn't include the minified code because I don't think it's particularly helpful, but I can include it if anybody wants me to.

As a general overview:

* Webpack is atrociously bad, in large part because each module has a huge amount of overhead (at least 1 function per module, plus a lot of extra stuff for import/export).

  In addition, the killer feature of Webpack 2 (dead code elimination) doesn't seem to be working. I suspect it's because Webpack 2 doesn't do any dead code elimination itself, it just relies on UglifyJS, but UglifyJS is a **very** weak optimizer.

  I think I should add in a comparison between Webpack 1 and Webpack 2. I suspect that Webpack 2 actually results in *more* code than Webpack 1.

  Very disappointing, but I know that they plan to improve the codegen for Webpack in the near future, so I'm slightly optimistic.

* `psc-bundle` is great, but it has overhead too (1 function per module, plus extra stuff like `PS` and importing/exporting). Unlike Webpack it has dead code elimination, so it does a lot better.

* In contrast, Rollup moves all of the code into a single big function and then renames variables to avoid conflicts. That means zero overhead, no functions, no importing, no exporting.

  And the Rollup dead code elimination is extremely good (especially considering it's designed for JavaScript!), I only found a few functions which should have been eliminated but weren't (which can hopefully be resolved with further pull requests to Rollup).

**Comment by Pauan** [comment]
Also, just to clarify, I'm using [this Rollup plugin created by @wegry and me](https://github.com/Pauan/purescript-tree-shaking/blob/master/rollup-plugin-purs.js) to compile the PureScript code with Rollup.

It's extraordinarily hacky, using regexps to convert CommonJS modules into ES6 modules. But it is possible to replace it with an actual proper plugin that does proper parsing (with Acorn or Esprima or whatever). That's my plan, anyways.

**Comment by wegry** [comment]
Thanks for that PR @Pauan. I have a bit of work on the transpiler/optimizer since then internally at work. I too want to move toward an AST based approach in some form. Especially for some of the peephole optimizations we've pegged on to it. `Data_Function.apply(f)(a)` is transformed into `(f)(a) /* Was Data_Function.apply */` and `(Data_Unit.unit)` becomes `(0 /* Was Data_Unit.unit */)`. Probably not good for a general bundling use. 😉 

We write the transpiled/optimized psc generated modules to disk with a node script prior to building in our minority native JS built with Webpack. Rollup could be socketed in in this place too (and probably should be from your benchmarks. It'll be quite a bit of work for us though.

**Comment by paf31** [comment]
@Pauan @wegry The output looks great! How much work do you think would be necessary to tidy up and package the plugin for general use?

If that could be done, I'd be interested in reducing the scope of `psc-bundle` before 1.0, and encouraging users to use Rollup or some other tool for bundling instead.

**Comment by wegry** [comment]
It will likely have to wait until after New Year's for me.

**Comment by paf31** [comment]
Oh sure, no rush, I was mostly just curious whether you think it's a lot of work or not.

**Comment by Pauan** [comment]
@paf31 I had some spare time over the weekend, so I started work on [a proper Rollup plugin for PureScript.](https://github.com/Pauan/rollup-plugin-purs)

However, I ran into an issue. It seems that PureScript sometimes imports/exports variables which are not valid JS identifiers.

As an example, `Data.Functor` exports this:

```
module.exports = {
    Functor: Functor, 
    flap: flap, 
    map: map, 
    mapFlipped: mapFlipped, 
    "void": $$void, 
    voidLeft: voidLeft, 
    voidRight: voidRight, 
    functorFn: functorFn, 
    functorArray: functorArray
};
```

The `"void"` part is the problem, because the Rollup plugin transforms the above code into this:

```
export {
    Functor,
    flap,
    map,
    mapFlipped,
    $$void as void,
    voidLeft,
    voidRight,
    functorFn,
    functorArray
 };
```

This transformation is necessary in order to take advantage of Rollup's tree shaking, but  `void` is not a valid identifier, so it breaks.

I can easily change the Rollup plugin so that it mangles the export, but mangling the imports is a **lot** harder, because it can't safely change `Data_Functor["void"]` to `Data_Functor.$$void` unless it keeps track of all the variables in all the scopes (so that it can determine which object lookups are module imports and which aren't).

So it's probably a lot easier to just change the `psc` output so that it mangles the imports/exports (which will be necessary eventually anyways with ES6 modules).

Should I file a separate issue for that?

**Comment by hdgarrood** [comment]
Does the spec say that you can't use reserved keywords as exports in ES6 modules?

**Comment by hdgarrood** [comment]
I was curious so I had a look, and according to the ES2015 spec you can use reserved words in exports declarations, so arguably I think this is a bug in Rollup and we shouldn't need to mangle names which are reserved words in JS.

**Comment by Pauan** [comment]
@hdgarrood You're right, using `export { foo as void };` is valid, and Rollup does accept that.

The situation is more complicated than that, however, because PureScript has re-exports. So you end up with situations like this:

```
var Data_Functor = require("../Data.Functor");

module.exports = {
  "void": Data_Functor["void"]
};
```

The Rollup plugin transforms that to this:

```
import * as Data_Functor from "../Data.Functor";

export var void = Data_Functor.void;
```

And now we're in trouble again. In order to fix that it would have to keep track of all variables in all scopes in order to know that `Data_Functor` is a module import, so that way it can rewrite the above to this:

```
export { void } from "./Data.Functor";
```

However, maybe I just need to bite the bullet and do the scope handling, even if it is difficult.

**Comment by wegry** [comment]
@pauan We encountered this problem ourselves, which spawned an issue that looks unrelated (and was shot down) issue I can't find at the moment talking about the compiler mangling reserved words from @paf31 and @ekmett. If a CJS export is an invalid ES identifier in <ES5 (delete) (or ES6+, const, await, async), we add the flag `purescript_[identifier]`.

All regex based, so not terribly robust. :P This of course is coming from Webpack 2. We've used it for two months and it works well. Not sure how it'll handle the bump from 0.9 to 0.10.

**Comment by Pauan** [comment]
@hdgarrood @paf31 Here's an update.

I was able to get the re-exporting to work without needing to do any scope analysis, but there is a [bug in Rollup](https://github.com/rollup/rollup/issues/1180) which causes re-exports to break with reserved words (importing/exporting works, it's only re-exporting that breaks).

So I'll try and get that bug fixed, and then hopefully soon I'll be able to re-run the tree shaking tests with the new plugin.

**Comment by Pauan** [comment]
@hdgarrood @paf31 I've made good progress, but I just now ran into another problem... some modules export identifiers which aren't valid JS identifiers at all:

```
// Data.Array/foreign.js
exports["uncons'"] = function (empty) {
  return function (next) {
    return function (xs) {
      return xs.length === 0 ? empty({}) : next(xs[0])(xs.slice(1));
    };
  };
};
```

```
// Data.Array/index.js
module.exports = {
    ...
    "group'": group$prime, 
    ...
};
```

Notice the `uncons'` and `group'` exports. This truly is invalid according to the ES spec. I can try adding in scope analysis and name mangling, but these really should be fixed on the PureScript side.

I think it would be a good idea if:

1. PureScript never generates exports which are invalid JS identifiers (like `group'`)

2. PureScript warns if an FFI file exports an invalid JS identifier (like `uncons'`)

In addition to improving compatibility with ES6 modules, this will also help with other backends (like C++), because other backends won't allow for the `'` character in imports/exports either.

I think FFI imports/exports should only use alphanumeric characters, for maximum compatibility with all backends.

Should I file a bug about this?

**Comment by paf31** [comment]
I would really like to be able to say that PS identifiers are just JS identifiers. That would mean removing primes, which I think would be a good thing, but it's been downvoted in the past.

**Comment by Pauan** [comment]
@paf31 There's nothing wrong with having primes in PureScript identifiers. It just means that the PureScript compiler needs to mangle primes, and that FFI files shouldn't be allowed to export primes.

**Comment by Pauan** [comment]
@paf31 For the record, I don't care whether primes are supported in PureScript or not. I'm okay with removing them, but I'm also okay with letting them stay. I'm just pointing out that there is a solution which solves this problem while also keeping primes.

**Comment by Pauan** [comment]
@paf31 As a temporary solution, I changed `rollup-plugin-purs` so that it mangles `'` to `$prime`

Here are the new results, using the new Rollup plugin:

* `psc-bundle` is `14.4 kB` regular, `6.8 kB` minified
* `rollup` is `9.0 kB` regular, `4.0 kB` minified
* `webpack` is `105.0 kB` regular, `23.7 kB` minified

I also updated [the comparison gist](https://gist.github.com/Pauan/3af525619a0bdf1f670043e0d70e2770) to use the latest results.

The reduction in size is mostly because the new plugin strips out some whitespace and comments, so there is no difference with the minified version.

**Comment by Pauan** [comment]
I got the [`rollup-plugin-purs`](https://github.com/Pauan/rollup-plugin-purs) plugin working, and I tried using it on [`tab-organizer`](https://github.com/Pauan/tab-organizer/tree/purescript), which is a PureScript application I've been developing.

It seems to work great:

* `webpack` is `579.7 kB` regular, `244.4 kB` minified
* `psc-bundle` is `165.1 kB` regular, `69.3 kB` minified
* `pulp-browserify` is `165.6 kB` regular, `69.1 kB` minified
* `rollup` is `113.8 kB` regular, `45.3 kB` minified

If you want to see the results for yourself, you can clone [this repository](https://github.com/Pauan/purescript-tree-shaking) and then follow the README instructions (you will need [yarn](https://yarnpkg.com/)).

If you want to use Rollup to bundle your own PureScript applications, you can start with [this configuration](https://github.com/Pauan/purescript-tree-shaking/blob/master/rollup.config.js) and then tweak it as needed. If anybody is interested, I can create a repository which demonstrates how to set up Rollup in more detail.

At this point it's mostly a matter of different people trying the plugin so that we can find all the bugs with it.

----

I did notice this warning:

```
output/Pauan.Animation/index.js (62:72) 'anyProxy' is not exported by 'output/Data.Generic/index.js'. See https://github.com/rollup/rollup/wiki/Troubleshooting#name-is-not-exported-by-module
```

It seems that `anyProxy` doesn't exist. I'm not sure if that's intentional or not, but it seems odd to me.

**Comment by paf31** [comment]
This is looking great! I'll try to use it over the next few weeks to test it out.

@Bodil @hdgarrood Do you think it makes sense to support this plugin in Pulp as an alternative to `psc-bundle`?

The `anyProxy` issue is documented here: https://github.com/purescript/purescript-generics/issues/44

**Comment by Pauan** [comment]
If you want to integrate Rollup into pulp, I recommend using [its JavaScript API](https://github.com/rollup/rollup/wiki/JavaScript-API). I can also give some suggestions on a good default configuration/plugins.

**Comment by hdgarrood** [comment]
I think it probably does make sense to support it in Pulp, although to be able to work out exactly how we would approach that I'd like to know more about how we're planning to reduce the scope of `psc-bundle`.

For example, is it possible to get Rollup to do something vaguely equivalent to `pulp build --to bundle.js --skip-entry-point --modules Foo,Bar`, producing a bundle which doesn't invoke any `main` function but instead provides a global variable `PS`, with properties `PS.Foo` and `PS.Bar` from which all exports of `Foo` and `Bar` are accessible? If not, is it feasible to keep this feature in `psc-bundle` but still delegate to Rollup for dead code elimination?

**Comment by Pauan** [comment]
@hdgarrood 
> For example, is it possible to get Rollup to do ...

* There is a [`rollup-plugin-multi-entry`](https://github.com/rollup/rollup-plugin-multi-entry) plugin which merges multiple entry points.

* It's trivial for me to change `rollup-plugin-purs` to provide an option to skip the `main` function call.

* You can provide a [`moduleName`](https://github.com/rollup/rollup/wiki/JavaScript-API#modulename) property which allows you to place the exports into an object (e.g. `PS`)

In general Rollup's plugin system gives a lot of power to do whatever you want with it.

**Comment by hdgarrood** [comment]
That all sounds great. So what are we imagining will be the scope of `psc-bundle` supposing we do start using Rollup for DCE instead?

**Comment by paf31** [comment]
Well the absolute minimum we need is something to replace the bundling step which happens in PSCi and Try PureScript. But that could be as simple as string concatenation and a `require` replacement, which could be much smaller and much faster than `psc-bundle`.

**Comment by Pauan** [comment]
@paf31 I haven't looked at the internals of PSCi or Try PureScript, but why do they need to use `psc-bundle`? Can't PSCi just evaluate the CommonJS files directly?

**Comment by wegry** [comment]
@pauan

The anyProxy bug was found by our Webpack implementation of ES module transpilation. As for your name mangling of prime, it is exactly what we did. 😉

**Comment by paf31** [comment]
@Pauan PSCi needs to bundle dependencies when using the browser backend, and Try PureScript for the same reason. It's possible that we could use some sort of `require` shim instead, but we would still need to bundle the sorted dependencies into a single file.

**Comment by Pauan** [comment]
@paf31 Ahh, I see. Rollup does have `dist/rollup.browser.js` which can be run in the browser. It's used on [the Rollup website](http://rollupjs.org/) to do browser-side bundling.

I'm not sure how easy it would be to integrate it into PSCi / Try PureScript, but it is another option. And if it worked then we could remove `psc-bundle` entirely.

**Comment by paf31** [comment]
Bundling needs to be done on the server in these cases really, for performance reasons. But we can still drastically reduce the scope of `psc-bundle`:

- The equivalent of `psc-bundle` would simply concatenate JS files and wrap them in a `require` shim if necessary
- The `psc-bundle` executable would go away
- Tools would have to switch to using Rollup via the plugin
- Any future optimizations would have to happen before codegen, ideally on the core so that they could be shared between backends.

I like this option, since `psc-bundle` is a bit of a hack right now, to say the least.

**Comment by Pauan** [comment]
I have now published my [`rollup-plugin-purs`](https://github.com/Pauan/rollup-plugin-purs) plugin to npm, so it's now possible to use `npm install rollup-plugin-purs` to install it.

I also added in an uncurrying optimization. All curried functions (global and local) will get a second uncurried version. If the curried function is called with all of its arguments, then it will use the uncurried version of the function instead.

Here are the new statistics:

* `rollup` (curried) is `113.9 kB` regular, `45.3 kB` minified
* `rollup` (uncurried) is `165.3 kB` regular, `45.4 kB` minified

The uncurried version is very slightly more code, but it probably runs faster, due to the lack of currying.

**Comment by paf31** [comment]
@Pauan Could you please post an example of the generated uncurried JS?

I'm very happy that it was possible to implement the uncurrying optimization outside `psc-bundle`. This gives me confidence that we can greatly reduce the scope of `psc-bundle` before 1.0 and push some optimizations into tools like `rollup-plugin-purs`.

Of course, we should measure the performance improvement before getting too excited, but this is great news!

**Comment by Pauan** [comment]
@paf31 This code:

```
var curriedVar = function (a) {
  return function (b) {
    return function (c) {
      return a + b + c;
    };
  };
};

function curriedFn(a) {
  return function (b) {
    return function (c) {
      return a + b + c;
    };
  };
}

curriedVar(1);
curriedVar(1)(2);
curriedVar(1)(2)(3);

curriedFn(1);
curriedFn(1)(2);
curriedFn(1)(2)(3);
```

Gets translated into this code:

```
var curriedVar__private_do_not_use_uncurried__3__0$0 = function (a, b, c) {
  return a + b + c;
};

var curriedFn__private_do_not_use_uncurried__3__0$0 = function (a, b, c) {
  return a + b + c;
};

var curriedVar = function (a) {
  return function (b) {
    return function (c) {
      return a + b + c;
    };
  };
};

function curriedFn(a) {
  return function (b) {
    return function (c) {
      return a + b + c;
    };
  };
}

curriedVar(1);
curriedVar(1)(2);
curriedVar__private_do_not_use_uncurried__3__0$0(1, 2, 3);

curriedFn(1);
curriedFn(1)(2);
curriedFn__private_do_not_use_uncurried__3__0$0(1, 2, 3);
```

It uses intentionally long function names to avoid conflicts, but minification shortens it to a single character.

If the function is never called fully saturated, then it won't generate the uncurried version.

Currently it only transforms curried functions which use a single function argument, so it won't change the `Fn*` functions, and it won't inline `Eff`.

There's still plenty of work to do, such as better heuristics for determining whether it's worth it to generate the uncurried function or not.

Personally, I think all optimizations should be in `psc`, because it has more information, and then it works with all backends, but this is a decent temporary alternative.

**Comment by Pauan** [comment]
I realized that the above compilation strategy can result in a huge amount of code duplication, so it now compiles to this instead:

```
var curriedVar__private_do_not_use_uncurried__3__0$0 = function (a, b, c) {
  return a + b + c;
};

var curriedFn__private_do_not_use_uncurried__3__0$0 = function (a, b, c) {
  return a + b + c;
};

var curriedVar = function (a) {
  return function (b) {
    return function (c) {
      return curriedVar__private_do_not_use_uncurried__3__0$0(a, b, c);
    };
  };
};

function curriedFn(a) {
  return function (b) {
    return function (c) {
      return curriedFn__private_do_not_use_uncurried__3__0$0(a, b, c);
    };
  };
}

curriedVar(1);
curriedVar(1)(2);
curriedVar__private_do_not_use_uncurried__3__0$0(1, 2, 3);

curriedFn(1);
curriedFn(1)(2);
curriedFn__private_do_not_use_uncurried__3__0$0(1, 2, 3);
```

Rather than duplicating the function body twice, it instead changes the curried function so that it calls the uncurried version.

Here are the new statistics:

* `rollup` (curried) is `113.9 kB` regular, `45.3 kB` minified
* `rollup` (uncurried) is `148.5 kB` regular, `47.8 kB` minified

**Comment by garyb** [comment]
I don't think that's a great idea, or at least it should be configurable - by doing that you're imposing a performance penalty on all uncurried code (which is probably the majority). If the file is served with gzip there will be minimal file size cost to duplicating the function bodies.

**Comment by wegry** [comment]
@garyb I think you mean curried code. I, for one, think that uncurried output should be the default and sweetened path. Duplicating code is not worth it, and doubling code size and potentially hampering startup time will not be alleviated by gzip or brotli. Binding exports that depend on curried functions are especially slow.

Making uncurried functons the default in output has the added benefit of making purescript exports idiomatically packagable as a lib for JS projects as well, instead of being hermetically sealed as it stands today.

**Comment by garyb** [comment]
Yeah, sorry, the curried code.

> I, for one, think that uncurried output should be the default and sweetened path

I don't have a problem with that, but given that it's the edge case in how code will be called in purescript I don't think imposing a global penalty on the usual case is worth it. I'm sure you don't agree though, which is why I suggested it should at least be a configurable option, that way everyone is happy.

**Comment by Pauan** [comment]
@garyb I understand your concerns, but the situation is more complicated than that:

* Modern JS JIT engines shouldn't have any problem at all with removing the overhead of the extra function call. Tracing JITs in particular are really good at that.

* Because modern JS JIT engines perform optimizations *per function*, duplicating function code means that the JIT has to duplicate optimizations too.

* Because JIT engines only optimize "hot" functions, if the function bodies are duplicated then there's a chance that the JIT won't optimize the functions at all, because the functions aren't called frequently enough.

* I suspect that JITs don't do a very good job optimizing curried functions, because they are extremely rare in ordinary JS code.

So it's entirely possible that the new code output is faster. Therefore we need extensive benchmarks to determine which compilation strategy is actually faster in practice.

Some interesting information: when compiling my decent-sized PureScript application, 177 calls were curried, and 85 were uncurried. However, most of the curried calls were because typeclass instances aren't being inlined. If typeclass instances were inlined, I think that would swing things in the favor of uncurried calls.

@wegry Unfortunately `rollup-plugin-purs` cannot be used to create more idiomatic JS packages from PureScript code. And the uncurrying optimization in particular is purely an optimization, so it doesn't affect the imports/exports. However, using idiomatic (uncurried) code does increase the chances that the JIT will be able to work its magic.

**Comment by garyb** [comment]
Trying https://jsperf.com/curried-vs-uncurried-body-overhead in

- MacOS Chrome
- MacOS Safari
- MacOS Firefox
- Windows Chrome
- Windows Firefox

The duplicate body case was faster in every single case, except for Safari where it was a tie. Invoking functions has an overhead, tracing VM or not.

**Comment by garyb** [comment]
> Some interesting information: when compiling my decent-sized PureScript application, 177 calls were curried, and 85 were uncurried. However, most of the curried calls were because typeclass instances aren't being inlined. If typeclass instances were inlined, I think that would swing things in the favor of uncurried calls.

I wanted to try the plugin on SlamData for a good real world example of how many curried vs uncurried functions there are, but unfortunately I wasn't able to. I spent about half an hour hacking case after case that it was complaining about but couldn't really afford to spend any more time on it. I'll open some issues when I get a chance. Mostly it was complaints about "invalid module" though, with a few cases of "duplicate export" (`toString` and `hasOwnProperty` are two cases of that that spring to mind, the former in `-argonaut`, the latter in `-foreign`).

**Comment by wegry** [comment]
@garyb My opinion about the priority of perf work [seems to be heterodox](https://github.com/purescript/purescript/issues/2267#issuecomment-238872097), and performance has historically been a sensitive issue. Please note I have no desire for a repeat of the tenor of #1206. That said, type-classes are indeed tremendously slow in our experience. The problem rears its head in JSON deserialization. We believed that [inlining](https://github.com/purescript/purescript/issues/2345) would save us a lot agony in that regard, but our pragma proposal looks to have died in committee.
 
The last time configuration came up, specifically concerning ES module exports and imports, [it was shot down as fragmenting the build chain](https://github.com/purescript/purescript/issues/1206#issuecomment-228821858). That portion could easily be in psc (and @texastoland had already done the work at one point). It's the whole reason this particular transpiler was started in the first place. And here we are with a JS based optimizer pass.

> I don't have a problem with that, but given that it's the edge case in how code will be called in purescript I don't think imposing a global penalty on the usual case is worth it.

Having blessed support of JS exports isn't a common use case because the blessed toolchain cannot currently support it. Both @pauan and I have been arguing for better JS support here and elsewhere because many real world projects are not green field and have to accommodate JS. At the very least, it would help with adoption of the language. 

And yes, if idiomatic exports aren't a thing still, I agree with you that the perf penalty of decurrying is not worth it.

The difference in uncurried v. curried perf at least in iOS Firefox was 3% (worth it in my opinion for idiomatic exports). On the desktop, I'm seeing a diff of 8%, which is significant. In the event that exports are not decurried, duplicating code or paying a >5% in runtime both seem to be unacceptable at least to me.

**Comment by Pauan** [comment]
@garyb Function calls have zero overhead when the function is optimized/traced/inlined/etc. [Benchmarking and performance is hard, especially with advanced JIT engines.](https://www.youtube.com/watch?v=g0ek4vV7nEA)

When I said "extensive benchmarking" I mean benchmarks of *actual programs*, not microbenchmarks (which are unreliable for predicting application performance).

Please do file bug reports.

**Comment by rvion** [comment]
I think I would slightly prefer de-duplication as a default, and code duplication (inlining) as an option:

some arguments I can think of are

   - easier to debug, manually tweak
   - easier to set breakpoints, etc
   - easier to generate source-maps
   - easier to profile: no randomly splited cost-centers
   - smaller file to generate/write/serve when developing
   - easier to parse output
   - clearer separation between currying related stuff and logic

> However, most of the curried calls were because typeclass instances aren't being inlined. If typeclass instances were inlined, I think that would swing things in the favor of uncurried calls.

That would really be great !

**Comment by paf31** [comment]
I don't have time to post a full reply right now, but let me just summarize: I'm not thinking about optimization yet, but I plan to think about it seriously after 1.0. Right now, I'm focusing on 1.0, and anything not already prioritized will be put in the "Approved" milestone unless there is a really good reason.

For 1.0, I would prefer that we cut DCE from `psc-bundle` if it can be replaced by external tools. I've created an issue to track that. But that's not to say that we can't return to DCE in the compiler later, if it can be done properly, and in a way which is compatible with ES6 without hacks.

I have no problem with supporting ES6 modules before 1.0, but I don't want to support two modes of code generation, and we would need to come up with a list of work to be done, since a change like that would mean changes to `psc`, Pulp, `purs-loader`, etc.

**Comment by michaelficarra** [comment]
@paf31 I just wanted to add that I think it would be a huge shame if we released 1.0 while generating CommonJS-based modules instead of real ECMAScript modules. In my opinion, it is an absolute must for the 1.0 release.

**Comment by paf31** [comment]
I agree, and if we cut features from `psc-bundle` at the same time, then there are no concerns with `language-javascript` any more, right?

**Comment by Pauan** [comment]
I would just like to point out that `rollup-plugin-purs` is an excellent way to transition toward ES6 modules: it works by converting CommonJS to ES6 modules, and it gives warnings if you use the dynamic features of CommonJS (which are incompatible with ES6 modules).

So by compiling your PureScript library/application with `rollup-plugin-purs` you'll be able to see which parts of your library are incompatible with ES6, that way you can fix those parts ahead of time.

In fact, it's possible to write a program which will take in CommonJS source code and spit back ES6 source code (which is what `rollup-plugin-purs` is already doing), which will help to automate the transition to ES6.

----

@paf31 From what I understand, the only reason for `language-javascript` is to ensure that the FFI files contain the appropriate exports, correct? But Rollup already verifies that the exports are correct (for all JS files), so I don't think it's necessary for `psc` to do those checks.

**Comment by hdgarrood** [comment]
@Pauan Inside `psc`, yes, that's the only reason we use `language-javascript`. Of course `psc-bundle` uses the AST for dead code elimination too, but I guess we already have an answer for that :)

I think I'd be happy to drop those checks from `psc`, provided that the errors that Rollup gives you in a case where you `foreign import` something which you forgot to define in the relevant FFI module are reasonably easy to understand.

**Comment by Pauan** [comment]
@hdgarrood Here is the error that `psc` gives:

```
Error found:

  The following values are not defined in the foreign module for module Pauan.View:

    mapImpl



See https://github.com/purescript/documentation/blob/master/errors/MissingFFIImplementations.md for more information,
or to contribute content related to this error.
```

Here is the warning that Rollup gives:

```
⚠️   'mapImpl' is not exported by 'output/Pauan.View/foreign.js'
https://github.com/rollup/rollup/wiki/Troubleshooting#name-is-not-exported-by-module
output/Pauan.View/index.js (20:52)
18:     return Data_Functor.map(Pauan_Stream.functorStream)(Pauan_StreamArray_Class.Replace.create)(Pauan_Stream.stream(toStreamView)(view1));
19: });
20: var functorView = new Data_Functor.Functor($foreign.mapImpl);
                                                        ^
21: var applyView = new Control_Apply.Apply(function () {
22:     return functorView;
```

It's a lot messier (because the message includes a few lines of actual source code), but it does tell you the name of the function (`mapImpl`), and which module it came from (`Pauan.View`)

**Comment by hdgarrood** [comment]
Nice :)

**Comment by Pauan** [comment]
I just pushed out version 1.0.4 of `rollup-plugin-purs`, which includes a `convert-commonjs` command which you can use to convert a CommonJS file to an ES6 module. You use it like this:

```
node_modules/.bin/convert-commonjs foo.js > foo.js
```

This will replace the CommonJS file `foo.js` with an ES6 file `foo.js`. This can be used to automate the conversion process from CommonJS to ES6 (but you might need to manually tweak the result). Make backups before you convert (Git counts as a backup).

**Comment by bodil** [comment]
While I'd obvs much rather see the compiler itself output ES6 natively, I'd say it's time to start playing around with integrating `rollup-plugin-purs` into Pulp. It could be a good place to de-complicate those JS specific error messages too.

**Comment by Pauan** [comment]
@bodil Of course, but there will be a transition period where PureScript libraries need to convert their FFI files from CommonJS to ES6, that's what `convert-commonjs` is for.

Also, as a side note, `rollup-plugin-purs` should work fine with either CommonJS or ES6, or even a mixture of different module systems.

**Comment by garyb** [comment]
@Pauan I've just opened a selection of issues that I encounter when trying to use the plugin on SlamData, although I've only just realised some of them are warnings only, and shouldn't block compilation (I think?). On Windows I just see squiggle-boxes rather than the icons or whatever it is that is supposed to be next to the errors, so it's not easy to tell which are fatal.

**Comment by Pauan** [comment]
@garyb Thanks, I'll address those now. If the compilation succeeded, then they were warnings, if the compilation failed then it was an error. Rollup seems to use `⚠️` for warnings and `🚨 ` for errors.

**Comment by garyb** [comment]
@Pauan that's why I can't tell, I never got the compilation to succeed :wink: and yeah, those do not show up as they're supposed to:

![squiggles](https://cloud.githubusercontent.com/assets/693642/21810573/488b0f00-d744-11e6-86fe-a0c93c38ecaa.png)

I'm assuming that it halts on a fatal error now though, so won't open issues for anything above the last message that I encounter in the future (unless the warning seems incorrect somehow).

**Comment by Pauan** [comment]
@garyb Ah yeah, I can see how that's confusing. I'm pretty sure that fatal errors cause it to immediately stop, so anything other than the bottom-most error is a warning.

**Comment by wegry** [comment]
@garyb Would you be able to paste the output that generates the dynamic require warning? Webpack has a similar constraint on static and verifiable imports (and exports). If `Debug.Trace` uses a CJS import and then attempts to modify it you could be in for a rough patch trying to move that over to ES Modules. This is something our code base ailed from moving from AMD and CJS to exclusively ES Modules. Your export names adversely affect this approach as well.

One of the major pitfalls of transpilation as a post processing step as opposed to being baked into psc is that name mangling is pretty tricky. With our internal transpilation method (which has been in use for some time now) we recently found a bug where exports that would be invalid ES6+ were affecting object properties of the same name (`default`). Because the current mangler (internally, and yet to be reconciled with @Pauan's pending go ahead from @gbaz) uses a regex to look for object property accessors it picks up plain Jane objects with that reserved key as well.

e.g.

`Made_Up_Module["default"]` becoming `Made_Up_Module.purescript_default` breaks on

```
var foo = {
  default: 'oh yeah'
}

foo.default
```
becoming 

```
var foo = {
  default: 'oh yeah'
}

foo.purescript_default
```

**Comment by garyb** [comment]
@wegry see the issues on the plugin repo. Those dynamic things are just warnings, so it's fine - they occur in a couple of modules when there is behaviour that needs patching so a module works in node as well as the browser.

**Comment by Pauan** [comment]
@wegry [This is the warning with the dynamic require](https://github.com/Pauan/rollup-plugin-purs/issues/5), but it's not actually a problem, it should work fine with `rollup-plugin-purs`

There is a [more serious issue with `virtual-dom`](https://github.com/Pauan/rollup-plugin-purs/issues/4), but `rollup-plugin-purs` was never intended to support arbitrary CommonJS modules, only PureScript code. I think the best way forward is to send a pull request to @Raynos which fixes the export issue. It's a really easy fix.

`rollup-plugin-purs` doesn't have any problems with importing/exporting reserved words. And it uses a full parse + AST (not regexps), so it doesn't mangle arbitrary objects, so that's not a problem either. In general `rollup-plugin-purs` should be quite solid.

**Comment by hdgarrood** [comment]
I think it might be time to close this and split it up into individual issues, as it's drifted a long way from the original proposal. We already have #2545; are there any other items that we need to track?

**Comment by Pauan** [comment]
@hdgarrood I created https://github.com/purescript/purescript/issues/2574 and https://github.com/purescript/purescript/issues/2575

**Comment by hdgarrood** [comment]
I think it's probably safe to assume that covers it then, so I'm going to close this. Please do add a comment if there's anything else I've missed.
